src/app/containers/LoginPage/index.js

```diff
@@ -20,47 +23,103 @@ import Loading from '../../common-ui/Loading';
+    <span>
+      Vui lòng liên hệ với fanpage tại địa chỉ{' '}
+      <a href="https://facebook.com/kc97blf">https://facebook.com/kc97blf</a> để được hỗ trợ
+    </span>,
+  ],
+};
+
+function validate(values) {
+  const newValues = {
+    ...values,
+    username: values.username || '',
+    password: values.password || '',
+  };
+  const errors = {
+    username: getUsernameErrorOrNull(newValues.username),
+    password: getPasswordErrorOrNull(newValues.password),
+  };
+  const hasError = Object.values(errors).some((error) => !!error);
+  return { newValues, errors, hasError };
+}
 
 function LoginPage({ history }) {
   const [apiProgress, setApiProgress] = React.useState(API_PROGRESS.INIT);
   const [showPopup, setShowPopup] = React.useState(false);
-  const [data, setData] = React.useState({
+  const [popupVariant, setPopupVariant] = React.useState(POPUP_VARIANT.INIT);
+  const [values, setValues] = React.useState({
     // null: pristine (user has not changed the value)
     // empty string: non-pristine (user has changed the value)
     username: null,
     password: null,
   });
+  const [errors, setErrors] = React.useState({});
   const { userInfo, setUserInfo } = React.useContext(UserInfoContext);
 
-  const handleChange = React.useCallback(
-    (name, value) => {
-      setData({ ...data, [name]: value });
-    },
-    [data],
-  );
-
   const handleSubmit = React.useCallback(
     async (event) => {
       event.preventDefault();
-      const { username, password } = data;
-      if (!username || !password) {
+      const { username, password } = values;
+      const validation = validate(values);
+      setValues(validation.newValues);
+      setErrors(validation.errors);
+      if (!username || !password || errors) {
         setShowPopup(true);
+        setPopupVariant(POPUP_VARIANT.ERROR);
         return;
       }
 
       setApiProgress(API_PROGRESS.REQ);
-      const { data: apiData } = await apiLogin({ username, password });
-      if (!apiData || !apiData.token) {
+      const { values: apivalues } = await apiLogin({ username, password });
+      if (!apivalues || !apivalues.token) {
         setShowPopup(true);
+        setPopupVariant(POPUP_VARIANT.ERROR);
         setApiProgress(API_PROGRESS.FAILED);
       } else {
-        setUserInfo({ ...userInfo, token: apiData.token });
+        setUserInfo({ ...userInfo, token: apivalues.token });
         history.push('/');
       }
     },
-    [data],
+    [values],
   );
 
+  const handleChange = (name, value) => {
+    setValues({ ...values, [name]: value });
+    setErrors({ ...errors, [name]: null });
+  };
+
+  const defaultProps = (name, value) => ({
+    label: labels[name],
+    name,
+    value,
+    onChange: (name, value) => handleChange(name, value),
+    error: errors[name],
+  });
+
+  const selectContent = (variant) => {
+    if (variant === POPUP_VARIANT.ERROR) return POPUP_MSG.error;
+    if (variant === POPUP_VARIANT.WARNING) return POPUP_MSG.warning[0];
```

Comment:
there are more than 2 errors, cannot use `variant` to determine content

Reply:
We might have more than 3 errors in the future, and `variant` only has 3 possible values. There might be two errors which use the same `variant`.

What should we use to determine the content?